internal/core/component_factory.rs

```diff
@@ -0,0 +1,287 @@
+use core::fmt::Debug;
+
+use alloc::{boxed::Box, rc::Rc};
+
+use crate::{
+    api::ComponentHandle,
+    component::{ComponentRc, ComponentVTable},
+};
+
+/// A `Component` built by a `ComponentFactory`
+#[repr(C)]
+pub struct Component(ComponentRc);
+
+impl<T: ComponentHandle> From<T> for Component
+where
+    T::Inner: vtable::HasStaticVTable<ComponentVTable> + 'static,
+{
+    fn from(value: T) -> Self {
+        Self(vtable::VRc::into_dyn(value.as_weak().inner.upgrade().unwrap()))
+    }
+}
+
+impl Component {
+    pub(crate) fn component_rc(self: Self) -> ComponentRc {
+        self.0
+    }
+
+    pub(crate) fn clone_internal(&self) -> Component {
+        Component(self.0.clone())
+    }
+}
+
+pub(crate) mod internal {
+    use super::{c_void, Component, Rc};
+
+    /// A internal helper needed to implement `ComponentFactory` in different languages.
+    #[derive(Clone, PartialEq)]
+    #[repr(C)]
+    pub struct ComponentFactoryData {
+        pub build: unsafe extern "C" fn(*mut c_void) -> *mut Component,
+        pub data: *mut c_void,
+        pub drop: unsafe extern "C" fn(*mut c_void),
+    }
+
+    #[allow(unsafe_code)]
+    impl ComponentFactoryData {
+        pub fn build(&self) -> Option<Component> {
+            let result = unsafe { (self.build)(self.data) };
+            if result.is_null() {
+                None
+            } else {
+                let r = unsafe { (*result).clone_internal() };
+                Some(r)
+            }
+        }
+    }
+
+    impl Default for ComponentFactoryData {
+        fn default() -> Self {
+            let data = Box::new(RustData { function: Box::new(|| None) });
+            Self {
+                build: __slint_component_factory_builder_call,
+                data: Box::into_raw(data) as *mut c_void,
+                drop: __slint_component_factory_builder_drop,
+            }
+        }
+    }
+
+    #[allow(unsafe_code)]
+    impl Drop for ComponentFactoryData {
+        fn drop(&mut self) {
+            unsafe { (self.drop)(self.data) };
+        }
+    }
+
+    pub(crate) struct RustData {
+        pub(crate) function: Box<dyn FnMut() -> Option<Component>>,
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn __slint_component_factory_builder_drop(data: *mut c_void) {
+        drop(Box::from_raw(data as *mut RustData));
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn __slint_component_factory_builder_call(
+        data: *mut c_void,
+    ) -> *mut Component {
+        let data = (data as *mut RustData).as_mut().unwrap();
+        if let Some(result) = (data.function)() {
+            Box::into_raw(Box::new(result))
+        } else {
+            core::ptr::null_mut()
+        }
+    }
+
+    pub(crate) struct ComponentFactoryInner(pub Rc<ComponentFactoryData>);
+}
+
+/// A Box containing a `ComponentFactory` (or nothing)
+#[repr(C)]
+pub struct ComponentFactory(core::ptr::NonNull<internal::ComponentFactoryInner>);
```

Comment:
Why this indirection?  Why not just `ComponentFactory(Rc<ComponentFactoryData>)`  (Or Option<Rc<_>>)
I know this is not repr(c), but we do just expose Rc anyway through FFI  (see the WindowAdapterRcOpaque)

The problem with this pointer is:
 1. one extra indirrection
 2. lots of unsafe code in this module

Another thing is that it looks like you are re-implementing what the vtable crate provides (via VBox) and maybe we can have an internal trait  (similar to the `ModelAdaptorVTable` in the interpreter)